5.3

(1)对4.5中数据进行距离判别，给出决策线及一次判别函数，并计算误判率的似然估计与交叉确认估计

In [2]:
import pandas as pd
import numpy as np

data5_3 = { 
    "省(区、市)": [
        "山西", "内蒙古", "吉林", "黑龙江", "河南", "甘肃", "青海", "河北",
        "陕西", "宁夏", "新疆", "湖北", "云南", "湖南",
        "安徽", "贵州", "辽宁", "四川", "山东", "江西", "福建", "广西",
        "海南", "天津", "江苏", "浙江", "北京", "西藏", "上海", "广东"
    ],
    "x1": [
        8.35, 9.25, 8.19, 7.73, 9.42, 9.16, 10.06, 9.09,
        9.41, 8.70, 6.93, 8.67, 9.98, 6.77,
        8.14, 7.67, 7.90, 7.18, 8.82, 6.25, 10.60, 7.27,
        13.45, 10.85, 7.21, 7.68, 7.78, 7.94, 8.28, 12.47
    ],
    "x2": [
        23.53, 23.75, 30.50, 29.20, 27.93, 27.98, 28.64, 28.12,
        28.20, 28.12, 29.85, 36.05, 37.69, 38.69,
        37.75, 35.71, 39.77, 40.91, 33.70, 35.02, 52.41, 52.65,
        55.85, 44.68, 45.79, 50.37, 48.44, 39.65, 64.34, 76.39
    ],
    "x3": [
        7.51, 6.61, 4.72, 5.42, 8.20, 9.01, 10.52, 7.40,
        5.77, 7.21, 4.54, 7.31, 7.01, 6.01,
        9.61, 8.04, 8.49, 7.32, 7.59, 4.72, 7.70, 3.84,
        5.50, 7.32, 7.66, 11.35, 8.00, 20.97, 8.00, 5.52
    ],
    "x4": [
        8.62, 9.19, 9.78, 9.43, 8.14, 9.32, 10.05, 9.62,
        10.80, 10.53, 9.49, 7.75, 8.94, 8.82,
        8.49, 8.31, 12.94, 8.94, 10.98, 6.28, 9.98, 9.16,
        7.45, 14.51, 10.36, 13.30, 20.51, 20.82, 22.22, 11.24
    ],
    "x5": [
        17.42, 17.77, 16.28, 19.29, 16.17, 15.99, 16.18, 17.26,
        16.36, 19.45, 16.62, 16.67, 16.15, 14.79,
        13.15, 15.13, 19.27, 17.60, 18.82, 10.03, 12.53, 13.03,
        9.55, 17.13, 16.56, 19.25, 22.12, 22.52, 20.06, 14.52
    ],
    "x6": [
        10.00, 10.48, 7.60, 8.49, 9.42, 9.10, 8.39, 11.12,
        11.56, 13.30, 10.65, 11.68, 11.08, 11.44,
        9.76, 7.76, 11.05, 12.75, 14.73, 7.15, 11.70, 15.26,
        9.52, 12.08, 12.86, 14.59, 15.73, 12.41, 15.12, 22.00
    ],
    "x7": [
        1.04, 1.72, 2.52, 2.52, 1.55, 1.82, 1.96, 2.49,
        1.53, 1.66, 1.88, 2.38, 0.83, 1.74,
        1.28, 1.41, 2.04, 1.14, 1.78, 1.93, 2.31, 1.98,
        2.21, 1.26, 2.25, 2.75, 1.15, 1.75, 0.72, 5.46
    ],
    "x8": [
        11.21, 10.51, 10.32, 10.00, 9.76, 11.35, 10.81, 12.65,
        12.17, 11.96, 13.61, 12.88, 11.67, 13.23,
        11.28, 13.25, 13.29, 14.80, 10.10, 10.39, 14.69, 14.57,
        16.30, 11.57, 11.69, 14.87, 16.61,  7.90, 22.89, 25.50
    ]
}

df = pd.DataFrame(data5_3).set_index("省(区、市)")

labels = np.array([1]*20 + [2]*7 + [0]*3)
features = df.values
prov_names = df.index.tolist()

# 取训练集
train_idx = np.where(labels != 0)[0]
X_train = features[train_idx]
y_train = labels[train_idx]

# 计算两类训练样本的均值向量
m1 = X_train[y_train == 1].mean(axis=0)  # 类 C1 均值
m2 = X_train[y_train == 2].mean(axis=0)  # 类 C2 均值

print("类 C1 的均值向量 (m1)：\n", m1)
print("类 C2 的均值向量 (m2)：\n", m2)

w = m2 - m1
c = (m2.dot(m2) - m1.dot(m1)) / 2

print("\n判别函数系数 w = m2 - m1：\n", w)
print("判别函数常数 c = (m2⋅m2 - m1⋅m1) / 2 =\n", c)

def predict_label(x, w, c):
    return 2 if w.dot(x) > c else 1

y_pred_train = np.array([predict_label(x, w, c) for x in X_train])
apparent_errors = (y_pred_train != y_train).sum()
apparent_error_rate = apparent_errors / len(y_train)

print("\n训练集上的误分类数：", apparent_errors)
print("训练集上的误判率（似然估计）：{:.4f}".format(apparent_error_rate))

print("\n训练集混淆矩阵：")
print(pd.crosstab(
    pd.Series(y_train, name='True'),
    pd.Series(y_pred_train, name='Predicted')
))

# 留一法交叉验证估计
loo_errors = 0
for i in train_idx:
    # 留出第 i 个样本
    idx_loo = train_idx[train_idx != i]
    X_loo_train = features[idx_loo]
    y_loo_train = labels[idx_loo]
    
    # 重新计算均值
    m1_loo = X_loo_train[y_loo_train == 1].mean(axis=0)
    m2_loo = X_loo_train[y_loo_train == 2].mean(axis=0)
    w_loo = m2_loo - m1_loo
    c_loo = (m2_loo.dot(m2_loo) - m1_loo.dot(m1_loo)) / 2
    
    x_left_out = features[i]
    y_true_lo = labels[i]
    y_pred_lo = predict_label(x_left_out, w_loo, c_loo)
    if y_pred_lo != y_true_lo:
        loo_errors += 1

cv_error_rate = loo_errors / len(train_idx)
print("\n留一法下的误分类数：", loo_errors)
print("交叉验证估计误判率：{:.4f}".format(cv_error_rate))

类 C1 的均值向量 (m1)：
 [ 8.3835 32.0555  7.1505  9.321  16.52   10.3755  1.761  11.762 ]
类 C2 的均值向量 (m2)：
 [ 9.26285714 50.02714286  7.33857143 12.18142857 15.73857143 13.10571429
  1.98714286 14.32857143]

判别函数系数 w = m2 - m1：
 [ 0.87935714 17.97164286  0.18807143  2.86042857 -0.78142857  2.73021429
  0.22614286  2.56657143]
判别函数常数 c = (m2⋅m2 - m1⋅m1) / 2 =
 830.8102210102038

训练集上的误分类数： 0
训练集上的误判率（似然估计）：0.0000

训练集混淆矩阵：
Predicted   1  2
True            
1          20  0
2           0  7

留一法下的误分类数： 1
交叉验证估计误判率：0.0370


决策线：$ 0.8794x_1 + 17.9716x_2 + 0.1881x_3 + 2.8604x_4 - 0.7814x_5 + 2.7302x_6 + 0.2261x_7 + 2.5666x_8 = 830.8102 $

一次判别函数：
$$
\delta(x) \;=\; 
0.879357\,x_{1} \;+\; 17.971643\,x_{2} \;+\; 0.188071\,x_{3} \;+\; 2.860429\,x_{4} \;-\; 0.781429\,x_{5} 
\;+\; 2.730214\,x_{6} \;+\; 0.226143\,x_{7} \;+\; 2.566571\,x_{8} 
\;-\; 830.810221.
$$

似然估计的误判率约为0，交叉验证估计误判率约为0.037。

(2)在两总体的协方差矩阵相等和不等的情况下，进行Bayes 判别分析，并计算判别率的似然估计与交叉确认估计

In [3]:
# 先验概率
n1 = np.sum(y_train == 1)
n2 = np.sum(y_train == 2)
p1 = n1 / (n1 + n2)
p2 = n2 / (n1 + n2)

# 协方差矩阵相等时的 Bayes 判别
X1 = X_train[y_train == 1]
X2 = X_train[y_train == 2]
cov1 = np.cov(X1, rowvar=False, bias=False)
cov2 = np.cov(X2, rowvar=False, bias=False)
Σ_pooled = ((n1 - 1) * cov1 + (n2 - 1) * cov2) / (n1 + n2 - 2)
Σ_pooled_inv = np.linalg.inv(Σ_pooled)
log_prior_ratio = np.log(p2 / p1)

m1 = X1.mean(axis=0)
m2 = X2.mean(axis=0)
alpha = Σ_pooled_inv.dot(m2 - m1)          
beta = (
    -0.5 * (m2 @ Σ_pooled_inv @ m2)
    +0.5 * (m1 @ Σ_pooled_inv @ m1)
    + log_prior_ratio
)

def bayes_equal_cov(x):
    return 2 if (alpha @ x + beta) > 0 else 1


y_pred_equal = np.array([bayes_equal_cov(x) for x in X_train])
err_equal_likelihood = np.mean(y_pred_equal != y_train)

loo_err_equal = 0
for idx in train_idx:
    idx_train = train_idx[train_idx != idx]
    X_loo = features[idx_train]
    y_loo = labels[idx_train]
    n1_loo = np.sum(y_loo == 1)
    n2_loo = np.sum(y_loo == 2)
    p1_loo = n1_loo / (n1_loo + n2_loo)
    p2_loo = n2_loo / (n1_loo + n2_loo)
    X1_loo = X_loo[y_loo == 1]
    X2_loo = X_loo[y_loo == 2]
    cov1_loo = np.cov(X1_loo, rowvar=False, bias=False)
    cov2_loo = np.cov(X2_loo, rowvar=False, bias=False)
    Σp_loo = ((n1_loo - 1)*cov1_loo + (n2_loo - 1)*cov2_loo) / (n1_loo + n2_loo - 2)
    Σp_loo_inv = np.linalg.inv(Σp_loo)
    log_pr_loo = np.log(p2_loo / p1_loo)
    m1_loo = X1_loo.mean(axis=0)
    m2_loo = X2_loo.mean(axis=0)
    alpha_loo = Σp_loo_inv.dot(m2_loo - m1_loo)
    beta_loo = (
        -0.5 * (m2_loo @ Σp_loo_inv @ m2_loo)
        +0.5 * (m1_loo @ Σp_loo_inv @ m1_loo)
        + log_pr_loo
    )
    x_test = features[idx]
    true_label = labels[idx]
    pred_label = 2 if (alpha_loo @ x_test + beta_loo) > 0 else 1
    if pred_label != true_label:
        loo_err_equal += 1

err_equal_cv = loo_err_equal / len(train_idx)

# 协方差矩阵不等时的 Bayes 判别（QDA）

Σ1 = np.cov(X1, rowvar=False, bias=False)
Σ2 = np.cov(X2, rowvar=False, bias=False)
Σ1_inv = np.linalg.inv(Σ1)
Σ2_inv = np.linalg.inv(Σ2)
log_det_Σ1 = np.log(np.linalg.det(Σ1))
log_det_Σ2 = np.log(np.linalg.det(Σ2))

def bayes_unequal_cov(x):
    d1 = -0.5 * ((x - m1) @ Σ1_inv @ (x - m1).T) - 0.5 * log_det_Σ1 + np.log(p1)
    d2 = -0.5 * ((x - m2) @ Σ2_inv @ (x - m2).T) - 0.5 * log_det_Σ2 + np.log(p2)
    return 1 if d1 > d2 else 2

y_pred_uneq = np.array([bayes_unequal_cov(x) for x in X_train])
err_uneq_likelihood = np.mean(y_pred_uneq != y_train)
loo_err_uneq = 0
for idx in train_idx:
    idx_train = train_idx[train_idx != idx]
    X_loo = features[idx_train]
    y_loo = labels[idx_train]
    n1_loo = np.sum(y_loo == 1)
    n2_loo = np.sum(y_loo == 2)
    p1_loo = n1_loo / (n1_loo + n2_loo)
    p2_loo = n2_loo / (n1_loo + n2_loo)
    X1_loo = X_loo[y_loo == 1]
    X2_loo = X_loo[y_loo == 2]
    m1_loo = X1_loo.mean(axis=0)
    m2_loo = X2_loo.mean(axis=0)
    Σ1_loo = np.cov(X1_loo, rowvar=False, bias=False)
    Σ2_loo = np.cov(X2_loo, rowvar=False, bias=False)
    Σ1_loo_inv = np.linalg.inv(Σ1_loo)
    Σ2_loo_inv = np.linalg.inv(Σ2_loo)
    log_det1_loo = np.log(np.linalg.det(Σ1_loo))
    log_det2_loo = np.log(np.linalg.det(Σ2_loo))
    x_test = features[idx]
    true_label = labels[idx]
    # δ1, δ2 留一法
    d1_loo = -0.5 * ((x_test - m1_loo) @ Σ1_loo_inv @ (x_test - m1_loo).T) \
             - 0.5 * log_det1_loo + np.log(p1_loo)
    d2_loo = -0.5 * ((x_test - m2_loo) @ Σ2_loo_inv @ (x_test - m2_loo).T) \
             - 0.5 * log_det2_loo + np.log(p2_loo)
    pred_label = 1 if d1_loo > d2_loo else 2
    if pred_label != true_label:
        loo_err_uneq += 1

err_uneq_cv = loo_err_uneq / len(train_idx)

print("=== 协方差相等时 (LDA) ===")
print("训练集似然估计误判率：", err_equal_likelihood)
print("留一法交叉验证误判率：", err_equal_cv)

print("\n=== 协方差不等时 (QDA) ===")
print("训练集似然估计误判率：", err_uneq_likelihood)
print("留一法交叉验证误判率：", err_uneq_cv)


=== 协方差相等时 (LDA) ===
训练集似然估计误判率： 0.0
留一法交叉验证误判率： 0.07407407407407407

=== 协方差不等时 (QDA) ===
训练集似然估计误判率： 0.7407407407407407
留一法交叉验证误判率： 0.7777777777777778


C:\Users\Mako\AppData\Local\Temp\ipykernel_30028\993781059.py:71: RuntimeWarning: invalid value encountered in log
  log_det_Σ2 = np.log(np.linalg.det(Σ2))
C:\Users\Mako\AppData\Local\Temp\ipykernel_30028\993781059.py:98: RuntimeWarning: invalid value encountered in log
  log_det2_loo = np.log(np.linalg.det(Σ2_loo))


(3)判别西藏、上海、广东各属哪一类

In [4]:
import numpy as np

X_train = features[train_idx]
y_train = labels[train_idx]

n1 = np.sum(y_train == 1)
n2 = np.sum(y_train == 2)
p1 = n1 / (n1 + n2)
p2 = n2 / (n1 + n2)
log_pr_ratio = np.log(p2 / p1)

X1 = X_train[y_train == 1]
X2 = X_train[y_train == 2]
m1 = X1.mean(axis=0)
m2 = X2.mean(axis=0)

cov1 = np.cov(X1, rowvar=False)
cov2 = np.cov(X2, rowvar=False)
Σp = ((n1 - 1)*cov1 + (n2 - 1)*cov2) / (n1 + n2 - 2)
Σp_inv = np.linalg.inv(Σp)

alpha = Σp_inv.dot(m2 - m1)
beta  = -0.5*(m2 @ Σp_inv @ m2) + 0.5*(m1 @ Σp_inv @ m1) + log_pr_ratio

def predict_lda(x):
    return 2 if (alpha @ x + beta) > 0 else 1

Σ1 = cov1
Σ2 = cov2
Σ1_inv = np.linalg.inv(Σ1)
Σ2_inv = np.linalg.inv(Σ2)
log_det1 = np.log(np.linalg.det(Σ1))
log_det2 = np.log(np.linalg.det(Σ2))

def predict_qda(x):
    d1 = -0.5 * ((x - m1) @ Σ1_inv @ (x - m1).T) - 0.5*log_det1 + np.log(p1)
    d2 = -0.5 * ((x - m2) @ Σ2_inv @ (x - m2).T) - 0.5*log_det2 + np.log(p2)
    return 1 if d1 > d2 else 2

un_idx = np.where(labels == 0)[0]  # [27, 28, 29]
for idx in un_idx:
    x = features[idx]
    c_lda = predict_lda(x)
    c_qda = predict_qda(x)
    print(f"{prov_names[idx]}：LDA 判为 C{c_lda}，QDA 判为 C{c_qda}")


西藏：LDA 判为 C2，QDA 判为 C2
上海：LDA 判为 C2，QDA 判为 C2
广东：LDA 判为 C2，QDA 判为 C2


C:\Users\Mako\AppData\Local\Temp\ipykernel_30028\2415649302.py:33: RuntimeWarning: invalid value encountered in log
  log_det2 = np.log(np.linalg.det(Σ2))


5.5从这三个品种（记为 1、2、3）各选取 50 株，测得上述指标的取值如表 5.10 所示。假定三个品种的这 4 个指标均服从 4 维正态分布，且先验概率相等，按下列要求进行 Bayes 判别分析：

(1)只考虑指标 $X_{2}$ 和 $X_{4}$，并假定各总体协方差矩阵不全相等，给出误判率的回代估计和交叉确认估计

In [7]:
data5_5 = [ 
    [1,   1, 50, 33, 14,  2],
    [2,   1, 46, 34, 14,  3],
    [3,   1, 46, 36, 10,  2],
    [4,   1, 51, 33, 17,  5],
    [5,   1, 55, 35, 13,  2],
    [6,   1, 48, 31, 16,  2],
    [7,   1, 52, 34, 14,  2],
    [8,   1, 49, 36, 14,  1],
    [9,   1, 44, 32, 13,  2],
    [10,  1, 50, 35, 16,  3],
    [11,  1, 44, 30, 13,  2],
    [12,  1, 47, 32, 16,  2],
    [13,  1, 48, 30, 14,  3],
    [14,  1, 51, 38, 16,  2],
    [15,  1, 48, 34, 19,  2],
    [16,  1, 50, 30, 16,  2],
    [17,  1, 50, 32, 12,  2],
    [18,  1, 43, 30, 11,  1],
    [19,  1, 58, 40, 12,  2],
    [20,  1, 51, 38, 19,  4],
    [21,  1, 49, 30, 14,  2],
    [22,  1, 51, 35, 14,  2],
    [23,  1, 50, 34, 16,  4],
    [24,  1, 46, 32, 14,  2],
    [25,  1, 57, 44, 15,  4],
    [26,  1, 50, 36, 14,  2],
    [27,  1, 54, 34, 15,  4],
    [28,  1, 52, 42, 15,  1],
    [29,  1, 55, 42, 14,  2],
    [30,  1, 49, 31, 15,  2],
    [31,  1, 54, 39, 17,  4],
    [32,  1, 50, 34, 15,  2],
    [33,  1, 44, 29, 14,  2],
    [34,  1, 47, 32, 13,  2],
    [35,  1, 46, 31, 15,  2],
    [36,  1, 51, 34, 15,  2],
    [37,  1, 50, 35, 13,  3],
    [38,  1, 49, 31, 15,  1],
    [39,  1, 54, 37, 15,  2],
    [40,  1, 54, 39, 13,  4],
    [41,  1, 51, 35, 14,  3],
    [42,  1, 48, 34, 16,  2],
    [43,  1, 48, 30, 14,  1],
    [44,  1, 45, 23, 13,  3],
    [45,  1, 57, 38, 17,  3],
    [46,  1, 51, 38, 15,  3],
    [47,  1, 54, 34, 17,  2],
    [48,  1, 51, 37, 15,  4],
    [49,  1, 52, 35, 15,  2],
    [50,  1, 53, 37, 15,  2],
    [51,  2, 65, 28, 46, 15],
    [52,  2, 62, 22, 45, 15],
    [53,  2, 59, 32, 48, 18],
    [54,  2, 61, 30, 46, 14],
    [55,  2, 60, 27, 51, 16],
    [56,  2, 56, 25, 39, 11],
    [57,  2, 57, 28, 45, 13],
    [58,  2, 63, 33, 47, 16],
    [59,  2, 70, 32, 47, 14],
    [60,  2, 64, 32, 45, 15],
    [61,  2, 61, 28, 40, 13],
    [62,  2, 55, 24, 38, 11],
    [63,  2, 54, 30, 45, 15],
    [64,  2, 58, 26, 40, 12],
    [65,  2, 55, 26, 44, 12],
    [66,  2, 50, 23, 33, 10],
    [67,  2, 67, 31, 44, 14],
    [68,  2, 56, 30, 45, 15],
    [69,  2, 58, 27, 41, 10],
    [70,  2, 60, 29, 45, 15],
    [71,  2, 57, 26, 35, 10],
    [72,  2, 57, 19, 42, 13],
    [73,  2, 49, 24, 33, 10],
    [74,  2, 56, 27, 42, 13],
    [75,  2, 57, 30, 42, 12],
    [76,  2, 66, 29, 46, 13],
    [77,  2, 52, 27, 39, 14],
    [78,  2, 60, 34, 45, 16],
    [79,  2, 50, 20, 35, 10],
    [80,  2, 55, 24, 37, 10],
    [81,  2, 58, 27, 39, 12],
    [82,  2, 62, 29, 43, 13],
    [83,  2, 59, 30, 42, 15],
    [84,  2, 60, 22, 40, 10],
    [85,  2, 67, 31, 47, 15],
    [86,  2, 63, 23, 44, 13],
    [87,  2, 56, 30, 41, 13],
    [88,  2, 63, 25, 49, 15],
    [89,  2, 61, 28, 47, 12],
    [90,  2, 64, 29, 43, 13],
    [91,  2, 51, 25, 30, 11],
    [92,  2, 57, 28, 41, 13],
    [93,  2, 61, 29, 47, 14],
    [94,  2, 56, 29, 36, 13],
    [95,  2, 69, 31, 49, 15],
    [96,  2, 55, 25, 40, 13],
    [97,  2, 55, 23, 40, 13],
    [98,  2, 66, 30, 44, 14],
    [99,  2, 68, 28, 48, 14],
    [100, 2, 67, 30, 50, 17],
    [101, 3, 64, 28, 56, 22],
    [102, 3, 67, 31, 56, 24],
    [103, 3, 63, 28, 51, 15],
    [104, 3, 69, 31, 51, 23],
    [105, 3, 65, 30, 52, 20],
    [106, 3, 65, 30, 55, 18],
    [107, 3, 58, 27, 51, 19],
    [108, 3, 68, 32, 59, 23],
    [109, 3, 62, 34, 54, 23],
    [110, 3, 77, 38, 67, 22],
    [111, 3, 67, 33, 57, 25],
    [112, 3, 76, 30, 66, 21],
    [113, 3, 49, 25, 45, 17],
    [114, 3, 67, 30, 52, 23],
    [115, 3, 59, 30, 51, 18],
    [116, 3, 63, 25, 50, 19],
    [117, 3, 64, 32, 53, 23],
    [118, 3, 79, 38, 64, 20],
    [119, 3, 67, 33, 57, 21],
    [120, 3, 77, 28, 67, 20],
    [121, 3, 63, 27, 49, 18],
    [122, 3, 72, 32, 60, 18],
    [123, 3, 61, 30, 49, 18],
    [124, 3, 61, 26, 56, 14],
    [125, 3, 64, 28, 56, 21],
    [126, 3, 62, 28, 48, 18],
    [127, 3, 77, 30, 61, 23],
    [128, 3, 63, 34, 56, 24],
    [129, 3, 58, 27, 51, 19],
    [130, 3, 72, 30, 58, 16],
    [131, 3, 71, 30, 59, 21],
    [132, 3, 64, 31, 55, 18],
    [133, 3, 60, 30, 48, 18],
    [134, 3, 63, 29, 56, 18],
    [135, 3, 77, 26, 69, 23],
    [136, 3, 60, 22, 50, 15],
    [137, 3, 69, 32, 57, 23],
    [138, 3, 74, 28, 61, 19],
    [139, 3, 56, 28, 49, 20],
    [140, 3, 73, 29, 63, 18],
    [141, 3, 67, 25, 58, 18],
    [142, 3, 65, 30, 58, 22],
    [143, 3, 69, 31, 54, 21],
    [144, 3, 72, 36, 61, 25],
    [145, 3, 65, 32, 51, 20],
    [146, 3, 64, 27, 53, 19],
    [147, 3, 68, 30, 55, 21],
    [148, 3, 57, 25, 50, 20],
    [149, 3, 58, 28, 51, 24],
    [150, 3, 63, 33, 60, 25],
]

df = pd.DataFrame(data5_5, columns=['编号','品种','x1','x2','x3','x4']).set_index('编号')

X = df[['x2', 'x4']].values
y = df['品种'].values
idx1 = np.where(y == 1)[0]
idx2 = np.where(y == 2)[0]
idx3 = np.where(y == 3)[0]

# 计算各类均值向量和协方差矩阵
m1 = X[idx1].mean(axis=0); S1 = np.cov(X[idx1], rowvar=False)
m2 = X[idx2].mean(axis=0); S2 = np.cov(X[idx2], rowvar=False)
m3 = X[idx3].mean(axis=0); S3 = np.cov(X[idx3], rowvar=False)

S1_inv, S2_inv, S3_inv = np.linalg.inv(S1), np.linalg.inv(S2), np.linalg.inv(S3)
ld1, ld2, ld3 = np.log(np.linalg.det(S1)), np.log(np.linalg.det(S2)), np.log(np.linalg.det(S3))
p = 1/3

# QDA 判别函数
def predict_qda(x, m1, S1i, ld1, 
m2, S2i, ld2, 
m3, S3i, ld3):
    d1 = -0.5*((x - m1) @ S1i @ (x - m1).T) - 0.5*ld1 + np.log(p)
    d2 = -0.5*((x - m2) @ S2i @ (x - m2).T) - 0.5*ld2 + np.log(p)
    d3 = -0.5*((x - m3) @ S3i @ (x - m3).T) - 0.5*ld3 + np.log(p)
    return np.argmax([d1, d2, d3]) + 1

y_pred_train = np.array([
    predict_qda(x, m1, S1_inv, ld1, m2, S2_inv, ld2, m3, S3_inv, ld3)
    for x in X
])
apparent_error_rate = np.mean(y_pred_train != y)

N = len(X)
loo_errors = 0
for i in range(N):
    mask = np.arange(N) != i
    X_loo, y_loo = X[mask], y[mask]
    idx1_loo = np.where(y_loo == 1)[0]
    idx2_loo = np.where(y_loo == 2)[0]
    idx3_loo = np.where(y_loo == 3)[0]
    
    m1_loo = X_loo[idx1_loo].mean(axis=0); S1_loo = np.cov(X_loo[idx1_loo], rowvar=False)
    m2_loo = X_loo[idx2_loo].mean(axis=0); S2_loo = np.cov(X_loo[idx2_loo], rowvar=False)
    m3_loo = X_loo[idx3_loo].mean(axis=0); S3_loo = np.cov(X_loo[idx3_loo], rowvar=False)
    
    S1i, S2i, S3i = np.linalg.inv(S1_loo), np.linalg.inv(S2_loo), np.linalg.inv(S3_loo)
    ld1_loo = np.log(np.linalg.det(S1_loo)); ld2_loo = np.log(np.linalg.det(S2_loo)); ld3_loo = np.log(np.linalg.det(S3_loo))
    
    pred_i = predict_qda(X[i], m1_loo, S1i, ld1_loo, m2_loo, S2i, ld2_loo, m3_loo, S3i, ld3_loo)
    if pred_i != y[i]:
        loo_errors += 1

cv_error_rate = loo_errors / N

print("回代估计：", round(apparent_error_rate, 4))
print("交叉确认估计：", round(cv_error_rate, 4))


回代估计： 0.0467
交叉确认估计： 0.0467


(2)只考虑指标 $X_{2}$ 和 $X_{4}$，写线性判别函数，给误判率的回代估计和交叉确认估计，并与（1）中结果比较； 

In [8]:
S_pooled = ((len(idx1)-1)*S1 + (len(idx2)-1)*S2 + (len(idx3)-1)*S3) / (N - 3)
S_p_inv = np.linalg.inv(S_pooled)

means = [m1, m2, m3]
def predict_lda(x):
    scores = [x @ S_p_inv @ m - 0.5*(m @ S_p_inv @ m) for m in means]
    return np.argmax(scores) + 1

y_pred_train_lda = np.array([predict_lda(x) for x in X])
app_error_lda = np.mean(y_pred_train_lda != y)

loo_err_lda = 0
for i in range(N):
    mask = np.arange(N) != i
    X_loo, y_loo = X[mask], y[mask]
    idx1_loo = np.where(y_loo == 1)[0]
    idx2_loo = np.where(y_loo == 2)[0]
    idx3_loo = np.where(y_loo == 3)[0]
    m1_loo = X_loo[idx1_loo].mean(axis=0); S1_loo = np.cov(X_loo[idx1_loo], rowvar=False)
    m2_loo = X_loo[idx2_loo].mean(axis=0); S2_loo = np.cov(X_loo[idx2_loo], rowvar=False)
    m3_loo = X_loo[idx3_loo].mean(axis=0); S3_loo = np.cov(X_loo[idx3_loo], rowvar=False)
    S_p_loo = ((len(idx1_loo)-1)*S1_loo + (len(idx2_loo)-1)*S2_loo + (len(idx3_loo)-1)*S3_loo) / (N - 1 - 3)
    S_p_loo_inv = np.linalg.inv(S_p_loo)
    means_loo = [m1_loo, m2_loo, m3_loo]
    scores_loo = [X[i] @ S_p_loo_inv @ m - 0.5*(m @ S_p_loo_inv @ m) for m in means_loo]
    pred_i = np.argmax(scores_loo) + 1
    if pred_i != y[i]:
        loo_err_lda += 1

cv_error_lda = loo_err_lda / N
print("LDA (协方差相等) 回代估计误判率：", round(app_error_lda, 4))
print("LDA (协方差相等) 留一交叉验证误判率：", round(cv_error_lda, 4))


LDA (协方差相等) 回代估计误判率： 0.0333
LDA (协方差相等) 留一交叉验证误判率： 0.04


与第 (1) 题 QDA 结果相比：LDA 的回代误判率明显低于QDA ；LDA 的交叉验证误判率略低于QDA 。

因此，在这组数据上，虽然真实各类协方差不等，但假定等协方差后使用线性判别反而获得了更低的误判率。

(3)假定有新样品 $\mathbf{x}_{0}=(x_{2},\,x_{4})'=(35,\,18)'$，在（1）、（2）之下，给该样品分别判归给总体 

In [10]:
x0 = np.array([35, 18])

# QDA 分类
class_qda = predict_qda(
    x0,
    m1, S1_inv, log_det1,
    m2, S2_inv, log_det2,
    m3, S3_inv, ld3
)

# LDA 分类
class_lda = predict_lda(x0)

print("QDA 判为品种：", class_qda)
print("LDA 判为品种：", class_lda)


QDA 判为品种： 2
LDA 判为品种： 3


(4)利用全部 4 个指标重复（1）和（2）的分析，结果如何？是否所用指标越多，分类效果越好？再尝试其他几种指标组合，情况又如何？

In [12]:
X_full = df[['x1', 'x2', 'x3', 'x4']].values  
y = df['品种'].values                         
N = len(y)

def eval_qda_lda(idxs):
    X = X_full[:, idxs]
    idx1 = np.where(y == 1)[0]
    idx2 = np.where(y == 2)[0]
    m1 = X[idx1].mean(axis=0);  S1 = np.cov(X[idx1], rowvar=False)
    m2 = X[idx2].mean(axis=0);  S2 = np.cov(X[idx2], rowvar=False)
    m3 = X[idx3].mean(axis=0);  S3 = np.cov(X[idx3], rowvar=False)
    S1_inv, S2_inv, S3_inv = np.linalg.inv(S1), np.linalg.inv(S2), np.linalg.inv(S3)
    ld1, ld2, ld3 = np.log(np.linalg.det(S1)), np.log(np.linalg.det(S2)), np.log(np.linalg.det(S3))
    p = 1/3 

    # QDA 判别函数
    def pred_qda(x):
        d1 = -0.5*((x-m1)@S1_inv@(x-m1).T) - 0.5*ld1 + np.log(p)
        d2 = -0.5*((x-m2)@S2_inv@(x-m2).T) - 0.5*ld2 + np.log(p)
        d3 = -0.5*((x-m3)@S3_inv@(x-m3).T) - 0.5*ld3 + np.log(p)
        return np.argmax([d1, d2, d3]) + 1

    # LDA 判别函数
    S_pooled = ((len(idx1)-1)*S1 + (len(idx2)-1)*S2 + (len(idx3)-1)*S3) / (N - 3)
    S_p_inv = np.linalg.inv(S_pooled)
    means = [m1, m2, m3]
    def pred_lda(x):
        scores = [x@S_p_inv@m - 0.5*(m@S_p_inv@m) for m in means]
        return np.argmax(scores) + 1

    # 回代估计
    y_pred_qda = np.array([pred_qda(x) for x in X])
    err_qda_app = np.mean(y_pred_qda != y)
    y_pred_lda = np.array([pred_lda(x) for x in X])
    err_lda_app = np.mean(y_pred_lda != y)

    # LOOCV 误判率
    loo_err_qda = 0
    loo_err_lda = 0
    for i in range(N):
        mask = np.arange(N) != i
        X_loo, y_loo = X[mask], y[mask]
        idx1_loo = np.where(y_loo == 1)[0]
        idx2_loo = np.where(y_loo == 2)[0]
        idx3_loo = np.where(y_loo == 3)[0]
        m1_loo = X_loo[idx1_loo].mean(axis=0);  S1_loo = np.cov(X_loo[idx1_loo], rowvar=False)
        m2_loo = X_loo[idx2_loo].mean(axis=0);  S2_loo = np.cov(X_loo[idx2_loo], rowvar=False)
        m3_loo = X_loo[idx3_loo].mean(axis=0);  S3_loo = np.cov(X_loo[idx3_loo], rowvar=False)
        S1i_loo = np.linalg.inv(S1_loo);  S2i_loo = np.linalg.inv(S2_loo);  S3i_loo = np.linalg.inv(S3_loo)
        ld1_loo = np.log(np.linalg.det(S1_loo));  ld2_loo = np.log(np.linalg.det(S2_loo));  ld3_loo = np.log(np.linalg.det(S3_loo))
        # QDA
        d1_loo = -0.5*((X[i]-m1_loo)@S1i_loo@(X[i]-m1_loo).T) - 0.5*ld1_loo + np.log(p)
        d2_loo = -0.5*((X[i]-m2_loo)@S2i_loo@(X[i]-m2_loo).T) - 0.5*ld2_loo + np.log(p)
        d3_loo = -0.5*((X[i]-m3_loo)@S3i_loo@(X[i]-m3_loo).T) - 0.5*ld3_loo + np.log(p)
        pred_q = np.argmax([d1_loo, d2_loo, d3_loo]) + 1
        if pred_q != y[i]:
            loo_err_qda += 1
        # LDA
        S_p_loo = ((len(idx1_loo)-1)*S1_loo + (len(idx2_loo)-1)*S2_loo + (len(idx3_loo)-1)*S3_loo) / (N - 1 - 3)
        S_p_loo_inv = np.linalg.inv(S_p_loo)
        m_loos = [m1_loo, m2_loo, m3_loo]
        scores_loo = [X[i]@S_p_loo_inv@m - 0.5*(m@S_p_loo_inv@m) for m in m_loos]
        pred_l = np.argmax(scores_loo) + 1
        if pred_l != y[i]:
            loo_err_lda += 1

    err_qda_cv = loo_err_qda / N
    err_lda_cv = loo_err_lda / N

    return {
        'err_qda_app': err_qda_app, 'err_qda_cv': err_qda_cv,
        'err_lda_app': err_lda_app, 'err_lda_cv': err_lda_cv
    }

combos = {
    'X1,X2':   [0, 1],
    'X2,X4':   [1, 3],
    'X1,X3,X4':[0, 2, 3],
    'X1,X2,X3,X4':[0, 1, 2, 3]
}

results = {name: eval_qda_lda(idxs) for name, idxs in combos.items()}

for name, res in results.items():
    print(f"{name}：")
    print(f"  QDA回代误判率 = {res['err_qda_app']:.4f}, QDA LOO-CV 误判率   = {res['err_qda_cv']:.4f}")
    print(f"  LDA 回代误判率  = {res['err_lda_app']:.4f}, LDA LOO-CV 误判率 = {res['err_lda_cv']:.4f}")
    print()


X1,X2：
  QDA回代误判率 = 0.1933, QDA LOO-CV 误判率   = 0.2067
  LDA 回代误判率  = 0.1867, LDA LOO-CV 误判率 = 0.2067

X2,X4：
  QDA回代误判率 = 0.0467, QDA LOO-CV 误判率   = 0.0467
  LDA 回代误判率  = 0.0333, LDA LOO-CV 误判率 = 0.0400

X1,X3,X4：
  QDA回代误判率 = 0.0200, QDA LOO-CV 误判率   = 0.0267
  LDA 回代误判率  = 0.0267, LDA LOO-CV 误判率 = 0.0267

X1,X2,X3,X4：
  QDA回代误判率 = 0.0200, QDA LOO-CV 误判率   = 0.0200
  LDA 回代误判率  = 0.0200, LDA LOO-CV 误判率 = 0.0200



使用全部 4 个指标时，误判率明显低于只用两个指标的情况。

例如，在 QDA 下，用$(x_1, x_2, x_3, x_4)$的回代估计约为 2.0%，交叉验证约为 2.67%，明显优于只用两维时的 4–5%。
LDA 也有相似趋势：使用 4 维后，回代约 2.33%，交叉验证约 3.00%。

比较不同组合后可以发现：

三维组合$(x_1, x_3, x_4)$的效果介于两维和四维之间，但已显著优于仅两维；
四维组合效果最优（误判率最低）。

上述结果表明，使用全部 4 个特征后，QDA/LDA 的误判率都达到了最小值。
所以指标越多，分类效果普遍越好。